### ISSUE TITLE
Synced Timezone wrong

### ISSUE THREAD

**Issue by Trefex** [inovex/CalendarSync#168]
Hi,

In my source, events are at 1 PM, but in the sink, they will be synced as 11 AM.

Are timezones perhaps not handled correctly or is it something i missed in a config?

Thanks
T

**Comment by alxndr13** [comment]
Hey! Thanks for the bug report! (and sorry for the late response!)

Could you give me some more details? Are you using different timezone's in both calendars?

**Comment by Trefex** [comment]
> Hey! Thanks for the bug report! (and sorry for the late response!)
> 
> Could you give me some more details? Are you using different timezone's in both calendars?

I basically sync to the same account in a different calendar for testing, so i guess the timezones are the same?

**Comment by Trefex** [comment]
@alxndr13 is it possible that in this function, timezone is not handled properly ? 

https://github.com/inovex/CalendarSync/blob/main/internal/adapter/outlook_http/client.go#L275

**Comment by alxndr13** [comment]
@Trefex i just found the time to dig deeper into this. 

Thanks for the bug report again! We definitely aren't handling timezones the way we should. I tested this as follows:

Created an event in my outlook **source** and made sure to use a different timezone. (Asia/Tomsk)

The event will then be represented in the API Response in UTC[1], the only fields which would cover the timezone correctly is the `originalStartTimeZone` and `originalEndTimeZone`[2]. The [DateTimeTimeZone](https://learn.microsoft.com/en-us/graph/api/resources/datetimetimezone?view=graph-rest-1.0) field just has the timezone set to UTC, which seems counterintuitive. 

[1]:
![screenshot 2024-09-04 at 19 11 25](https://github.com/user-attachments/assets/579be1b0-3c01-4a08-9f28-5df84ddb78af)

[2]: 
![screenshot 2024-09-04 at 19 10 53](https://github.com/user-attachments/assets/4d55e629-542b-4c3d-8317-c40bb0cd7d2b)

This is also why it is working all the time for my case, where i have outlook as the sink: i just retrieve the UTC Time and write it in UTC to my google sink. The clients then represent the events according to my machines timezone. 

In case we use outlook as a **sink**, we should make sure to convert (and set) the timezone to UTC when retrieving events. Your local client should then show the time according to your preferred timezone. 

Need to attend a padel match now, will have a deeper look at this in the coming days i hope.

**Comment by Trefex** [comment]
@alxndr13 this is indeed documented here it seems: https://learn.microsoft.com/en-us/graph/api/user-list-events?view=graph-rest-1.0&tabs=http#support-various-time-zones

**Comment by alxndr13** [comment]
Hi @Trefex,

could you try to build the app with the `fix-168` branch and test if that fixes your problem?

**Comment by Trefex** [comment]
@alxndr13 it is the same result. All events are shifted by 2 hours (I am in CEST).

From the PR, it seems though as the timezone is not set / added in https://github.com/inovex/CalendarSync/blob/main/internal/adapter/outlook_http/client.go#L275 ?

I think the timeZone need to be set as well when writing back the event in Outlook, otherwise you will write the time in UTC?

```
➜  bin git:(fix-168) ./calendarsync --version
Version: 7e7d652
```

**Comment by alxndr13** [comment]
ah, understood. This is uncoventional, as this differentiates it from the google adapter.. When writing calendar events into the google sink using UTC as the events timezone, it is still always represented in your computers timezone. (which makes sense, it's the same time, just differently displayed)

Example flow when outlook is my source and google my sink: 

- Event is retrieved from outlook, time is 3pm CEST
- gets converted into our internal format and to UTC (1pm)
- written to google in UTC (1pm)
- but gets represented in my browser as 3pm 

Which is what i would expect.. 

But outlook seems to handle this differently and presents the event in it's original timezone and not your local timezone, which doesnt make sense imo. (why would you look at events where each event has a different timezone??) 

(not sure if this is actually the case but this would explain the behaviour it seems)

**Comment by Trefex** [comment]
@alxndr13 what if you use Outlook as source and Outlook as sync ? I can provide a calendar JSON if that would help.

**Comment by alxndr13** [comment]
@Trefex used google as source, outlook as sink in the example above. 

Calendar JSOn would help, yes. Please make sure to remove all personal information

**Comment by Trefex** [comment]
Source (Outlook)

```{
    "@odata.context": "xxx",
    "@odata.etag": "xxxx",
    "id": "xxxx",
    "subject": "xxxx",
    "bodyPreview": "Rolling minutes: xxx",
    "body": {
        "contentType": "html",
        "content": "xxxx"
    },
    "start": {
        "dateTime": "2024-09-20T11:00:00.0000000",
        "timeZone": "UTC"
    },
    "end": {
        "dateTime": "2024-09-20T12:00:00.0000000",
        "timeZone": "UTC"
    },
    "location": {
        "displayName": "xxxx",
        "locationType": "default",
        "uniqueId": "xxxxx",
        "uniqueIdType": "private"
    },
    "attendees": [
        {
            "type": "required",
            "status": {
                "response": "none",
                "time": "0001-01-01T00:00:00Z"
            },
            "emailAddress": {
                "name": "xxxx",
                "address": "xxxx"
            }
        },
        {
            "type": "required",
            "status": {
                "response": "none",
                "time": "0001-01-01T00:00:00Z"
            },
            "emailAddress": {
                "name": "xxxx",
                "address": "xxxx"
            }
        }
    ],
    "organizer": {
        "emailAddress": {
            "name": "xxxx",
            "address": "xxxxx"
        }
    },
    "calendar@odata.associationLink": "xxxx",
    "calendar@odata.navigationLink": "xxxx"
}
``` 

Sink

```

```
    "value": [
        {
            "@odata.etag": "xxxx",
            "id": "xxxx",
            "createdDateTime": "2024-09-20T06:52:01.9038305Z",
            "lastModifiedDateTime": "2024-09-20T06:54:03.2007083Z",
            "changeKey": "xxxx",
            "categories": [],
            "transactionId": null,
            "originalStartTimeZone": "Europe/Luxembourg",
            "originalEndTimeZone": "Europe/Luxembourg",
            "iCalUId": "xxxx",
            "uid": "xxxxx",
            "reminderMinutesBeforeStart": 10,
            "isReminderOn": true,
            "hasAttachments": false,
            "subject": "[ELU] xxxx",
            "bodyPreview": "xxxx",
            "importance": "normal",
            "sensitivity": "normal",
            "isAllDay": false,
            "isCancelled": false,
            "isOrganizer": true,
            "responseRequested": true,
            "seriesMasterId": null,
            "showAs": "busy",
            "type": "singleInstance",
            "webLink": "xxxx",
            "onlineMeetingUrl": null,
            "isOnlineMeeting": false,
            "onlineMeetingProvider": "unknown",
            "allowNewTimeProposals": true,
            "occurrenceId": null,
            "isDraft": false,
            "hideAttendees": false,
            "responseStatus": {
                "response": "organizer",
                "time": "0001-01-01T00:00:00Z"
            },
            "body": {
                "contentType": "html",
                "content": "xxxx"
            },
            "start": {
                "dateTime": "2024-09-04T08:00:00.0000000",
                "timeZone": "UTC"
            },
            "end": {
                "dateTime": "2024-09-04T09:00:00.0000000",
                "timeZone": "UTC"
            },
            "location": {
                "displayName": "xxxx",
                "locationType": "default",
                "uniqueId": "xxxx",
                "uniqueIdType": "private"
            },
            "locations": [
                {
                    "displayName": "xxxx",
                    "locationType": "default",
                    "uniqueId": "xxxxx",
                    "uniqueIdType": "private"
                }
            ],
            "recurrence": null,
            "attendees": [],
            "organizer": {
                "emailAddress": {
                    "name": "xxxx",
                    "address": "xxxxx"
                }
            },
            "onlineMeeting": null
        },
```

**Comment by alxndr13** [comment]
Hi @Trefex,

fyi: Yesterday was my last day at Inovex. Since I won't be using CalendarSync daily anymore, addressing this issue will take some time as I'm focusing on my new role.

**Comment by MichaelEischer** [comment]
I've read through the issue and am still confused what's going on. 

https://github.com/inovex/CalendarSync/blob/ccdebc534f875ae0c3a6aa2fd3ee185a553e3ea8/internal/adapter/outlook_http/client.go#L186 looks suspicious though. It determines the local TZ and sets that for the timestamps of a to be written event. That could explain what you're seeing. The UTC timestamp `2024-09-20T11:00:00` gets marked as CEST. After normalizing that timestamp again to UTC, we're now at `2024-09-20T9:00:00`. (not sure why the event snippets show a time diff of three hours though).

Can you try whether setting the environment variable `TZ=UTC` makes the problem disappear?

**Comment by Trefex** [comment]
@MichaelEischer this indeed fixes the issue. I have now tested on the #187 branch.

Would this require a fix in the code or should the docs be updated to reflect the TZ env variable be set to UTC

Thank you !

**Comment by MichaelEischer** [comment]
> Would this require a fix in the code

This definitely should be fixed in the code, the current inconsistent timezone handling is definitely broken. Not sure when I'll get around to fix it.

**Comment by Trefex** [comment]
No worries, in the meantime, the workaround is fine.

I am more curious about #166 :)

**Comment by MichaelEischer** [comment]
fixed via #187.
